lib/hotel_block.rb

```diff
@@ -0,0 +1,96 @@
+require 'date'
+# # require_relative 'hotel.rb'
+
+module Booking
+  class Block
+
+    attr_reader :checkin_date, :checkout_date, :requested_rooms, :discounted_rate, :party_name, :room_block, :block_of_rooms, :all_rooms, :max_rooms, :room_numbers, :hotel, :all_reservations
+    
+    def initialize(checkin_date, checkout_date, requested_rooms, discounted_rate, hotel: nil)
+    @checkin_date = checkin_date
+    @checkout_date = checkout_date 
+    @block_of_rooms ||= []
+    @hotel ||= hotel
+    @available_rooms = @hotel.check_availability(checkin_date, checkout_date)
+    @all_reservations = @hotel.all_reservations
+    @discounted_rate = discounted_rate
+
+    requested_rooms.each do |room|
+      if @block_of_rooms.include?(room)
+        raise ArgumentError, "That room is already in a block."
+      elsif !@available_rooms.include?(room)
+        raise ArgumentError, "That room is unavailable"
+      else
+        @requested_rooms = requested_rooms
+      end
+    end
+  
+    @room_block = []
+    @room_numbers = room_numbers
+    @all_rooms = @hotel.all_rooms
+  end
+
+    # Method to ensure that the requested room numbers are valid
+    def room_numbers
+      i = 0
+
+      requested_rooms.each do |room|
+        unless @available_rooms == nil
+          if !@available_rooms.include?(room)
+            raise ArgumentError, "That room is unavailable."
+          end
+        end
+
+        if requested_rooms.length > 5
```

Comment:
This logic for checking the length of the `requested_rooms` is inside of an each loop, looking at each of the requested rooms. This conditional could happen outside of the each loop... before the each loop... and probably belongs outside of this method and up in the `initialize` method! This check should only happen once, and I would put it next to all of the other places where you check that this `Block` has valid parameters